File: app/src/main/java/com/cameronbanga/skyscraper/services/AppTheme.kt
package com.cameronbanga.skyscraper.services

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.toArgb
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Manages app-wide theme colors
 */
class AppTheme private constructor(private val context: Context) {

    companion object {
        @Volatile
        private var INSTANCE: AppTheme? = null

        fun getInstance(context: Context): AppTheme = INSTANCE ?: synchronized(this) {
            INSTANCE ?: AppTheme(context.applicationContext).also { INSTANCE = it }
        }

        val shared get() = INSTANCE
            ?: throw IllegalStateException("AppTheme not initialized")
    }

    private val prefs: SharedPreferences = context.getSharedPreferences(
        "skyscraper_theme",
        Context.MODE_PRIVATE
    )

    private val _accentColor = MutableStateFlow(ThemeColor.Orange.color)
    val accentColor: StateFlow<Color> = _accentColor.asStateFlow()

    private val _isCustomColor = MutableStateFlow(false)
    val isCustomColor: StateFlow<Boolean> = _isCustomColor.asStateFlow()

    init {
        loadColor()
    }

    private fun loadColor() {
        val colorName = prefs.getString("accent_color", null)

        when {
            colorName == "custom" -> {
                _isCustomColor.value = true
                _accentColor.value = loadCustomColor() ?: ThemeColor.Orange.color
            }
            colorName != null -> {
                val themeColor = ThemeColor.values().firstOrNull { it.name == colorName }
                _isCustomColor.value = false
                _accentColor.value = themeColor?.color ?: ThemeColor.Orange.color
            }
            else -> {
                _isCustomColor.value = false
                _accentColor.value = ThemeColor.Orange.color
            }
        }
    }

    fun setThemeColor(themeColor: ThemeColor) {
        _isCustomColor.value = false
        _accentColor.value = themeColor.color
        saveColor(themeColor.name)
    }

    fun setCustomColor(color: Color) {
        _isCustomColor.value = true
        _accentColor.value = color
        saveCustomColor(color)
    }

    private fun saveColor(colorName: String) {
        prefs.edit()
            .putString("accent_color", colorName)
            .apply()

        // Clear custom color data
        prefs.edit()
            .remove("custom_color_argb")
            .apply()
    }

    private fun saveCustomColor(color: Color) {
        prefs.edit()
            .putString("accent_color", "custom")
            .putInt("custom_color_argb", color.toArgb())
            .apply()
    }

    private fun loadCustomColor(): Color? {
        val argb = prefs.getInt("custom_color_argb", 0)
        return if (argb != 0) {
            Color(argb)
        } else {
            null
        }
    }
}

enum class ThemeColor(val displayName: String, val color: Color) {
    Orange("Orange", Color(0xFFFF9500)),
    Blue("Blue", Color(0xFF007AFF)),
    Purple("Purple", Color(0xFFAF52DE)),
    Pink("Pink", Color(0xFFFF2D55)),
    Green("Green", Color(0xFF34C759)),
    Red("Red", Color(0xFFFF3B30)),
    Teal("Teal", Color(0xFF5AC8FA)),
    Indigo("Indigo", Color(0xFF5856D6)),
    Yellow("Yellow", Color(0xFFFFCC00))
}

File: Skyscraper-iOS-20251106/Skyscraper/Services/AppTheme.swift
//
//  AppTheme.swift
//  Skyscraper
//
//  Manages app-wide theme colors
//

import SwiftUI
import Combine

@MainActor
class AppTheme: ObservableObject {
    static let shared = AppTheme()

    @Published var accentColor: Color {
        didSet {
            saveColor()
        }
    }

    @Published var isCustomColor: Bool = false

    private init() {
        // Initialize with default first
        self.accentColor = ThemeColor.orange.color

        // Load saved color or keep default
        if let colorName = UserDefaults.standard.string(forKey: "accentColor") {
            if colorName == "custom" {
                // Load custom color from UserDefaults
                self.isCustomColor = true
                self.accentColor = loadCustomColor() ?? ThemeColor.orange.color
            } else if let color = ThemeColor.allCases.first(where: { $0.name == colorName }) {
                self.isCustomColor = false
                self.accentColor = color.color
            } else {
                self.isCustomColor = false
                self.accentColor = ThemeColor.orange.color
            }
        } else {
            self.isCustomColor = false
            self.accentColor = ThemeColor.orange.color
        }
    }

    private func saveColor() {
        if isCustomColor {
            // Save as custom color
            UserDefaults.standard.set("custom", forKey: "accentColor")
            saveCustomColor(accentColor)
        } else if let themeColor = ThemeColor.allCases.first(where: { $0.color == accentColor }) {
            // Save predefined color
            UserDefaults.standard.set(themeColor.name, forKey: "accentColor")
            // Clear custom color data
            UserDefaults.standard.removeObject(forKey: "customColorRed")
            UserDefaults.standard.removeObject(forKey: "customColorGreen")
            UserDefaults.standard.removeObject(forKey: "customColorBlue")
            UserDefaults.standard.removeObject(forKey: "customColorOpacity")
        }
    }

    func setCustomColor(_ color: Color) {
        isCustomColor = true
        accentColor = color
    }

    private func saveCustomColor(_ color: Color) {
        #if os(iOS)
        let uiColor = UIColor(color)
        var red: CGFloat = 0
        var green: CGFloat = 0
        var blue: CGFloat = 0
        var alpha: CGFloat = 0

        uiColor.getRed(&red, green: &green, blue: &blue, alpha: &alpha)

        UserDefaults.standard.set(Double(red), forKey: "customColorRed")
        UserDefaults.standard.set(Double(green), forKey: "customColorGreen")
        UserDefaults.standard.set(Double(blue), forKey: "customColorBlue")
        UserDefaults.standard.set(Double(alpha), forKey: "customColorOpacity")
        #endif
    }

    private func loadCustomColor() -> Color? {
        guard UserDefaults.standard.object(forKey: "customColorRed") != nil else {
            return nil
        }

        let red = UserDefaults.standard.double(forKey: "customColorRed")
        let green = UserDefaults.standard.double(forKey: "customColorGreen")
        let blue = UserDefaults.standard.double(forKey: "customColorBlue")
        let opacity = UserDefaults.standard.double(forKey: "customColorOpacity")

        return Color(red: red, green: green, blue: blue, opacity: opacity)
    }
}

enum ThemeColor: String, CaseIterable, Identifiable {
    case orange
    case blue
    case purple
    case pink
    case green
    case red
    case teal
    case indigo
    case yellow

    var id: String { rawValue }

    var name: String {
        rawValue.capitalized
    }

    var color: Color {
        switch self {
        case .orange: return .orange
        case .blue: return .blue
        case .purple: return .purple
        case .pink: return .pink
        case .green: return .green
        case .red: return .red
        case .teal: return .teal
        case .indigo: return .indigo
        case .yellow: return .yellow
        }
    }
}
